File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from collections import deque

n, k = map(int, input().split())
visited = [0] * 100001
q = deque()
q.append(n)
visited[n] = 1
answer = 0

while q:
    value = q.popleft()
    if value == k:
        break

    for now in (value * 2, value-1, value + 1):
        if now < 0 or now > 100000:
            continue
        if visited[now]:
            continue

        if now == value * 2:
            visited[now] = visited[value]
        else:
            visited[now] = visited[value] + 1
        q.append(now)
print(visited[k]-1)

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "ex.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [n, k] = input[0].split(" ").map(Number);
const visited = Array(100001).fill(0);

const bfs = () => {
  q = [];
  q.push([n, 0]);
  visited[n] = 1;

  while (q.length) {
    const [now, cnt] = q.shift();

    if (now === k) {
      console.log(cnt);
      return;
    }
    for (const nx of [now * 2, now - 1, now + 1]) {
      if (nx >= 0 && nx <= 100000 && visited[nx] === 0) {
        if (nx === now * 2) {
          q.push([nx, cnt]);
        } else {
          q.push([nx, cnt + 1]);
        }
        visited[nx] = 1;
      }
    }
  }
};

bfs();
